### DISCUSSION TITLE
convert creates incomplete PDFs

### DISCUSSION THREAD

**Discussion by YutMarma** [ImageMagick/ImageMagick#7172] [Help]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

Ubuntu 23.10

### Description

I have 38 .PNGs that I'd like to turn into a single .PDF.

`convert *.png test.pdf` outputs a .pdf with no error messages. However, the .PDF only includes the first 11 images.

Online converters, such as https://png2pdf.com/, handle things fine, so I don't believe there's an issue with any of the .PNGs themselves.

Obviously, I'd rather not have to upload 40+ .PNGs to an online converter, which is why I'm hoping I can get ImageMagick working.

Any suggested troubleshooting steps? I'm unable to update to version 7.x on Ubuntu using apt, I tried downloading the .rpms from the official site but even after running `alien` there were still a ton of missing dependencies, so unless and until the "famously user-friendly" Linux distro can catch up to Microsoft Windows XP's ease of application installation, I'm stuck on 6.9.11-60.

**Comment by fmw42** [comment]
Add -adjoin to your command after *.png

**Reply by YutMarma** [reply]
What an incredibly speedy reply. Thank you!

```
user@MACHINE:~/idkme$ convert --help | grep adjoin
  -adjoin              join images into a single multi-image file
```

Out of curiosity - is there a brief technical explanation for why some of the files joined even without the `-adjoin` argument?

**Reply by YutMarma** [reply]
Whoops, looks like I spoke too soon - the command seemed to be taking an extra while to run, so I assumed it was working, but instead I just have the same 11-page version I was getting without `-adjoin`. 

I guess that at least explains why some were joining without the command - `convert` already knew enough to silently assume it.

**Reply by fmw42** [reply]
Sorry, did you want multiple outputs?  If so, change that to +adjoin.

Some formats such as pdf, gif, tiff support multiple pages (by default). So -adjoin is not needed. But +adjoin would be if you want separate pages.  Other formats do not support multiple pages and would always create multiple outputs even if specifying -adjoin.

**Reply by YutMarma** [reply]
No, you had it right the first time! I'm looking for one 38-page .pdf with the contents of all 38 .png files.

Interestingly, `convert *.png +adjoin test.pdf` ends up creating 11 separate .pdfs. I don't know what it is about the 12th .png that's making things go haywire!

**Comment by snibgo** [comment]
I suspect there is something weird about the 12th file.

What does `convert *.png info:` say? What about `convert *.png -verbose info:`?

**Reply by YutMarma** [reply]
> I suspect there is something weird about the 12th file.

So would I, except it works fine in an online converter, and all files were generated as part of the same batch process. I even went back and re-created them from the original .TIFs, with no change.

> What does `convert *.png info:` say?

```
user@COMP:~/idkme$ convert *.png info:
ScanImage329_1L.png[0] PNG 3514x2950 3514x2950+0+0 8-bit sRGB 256c 5.87618MiB 1.470u 0:01.229
ScanImage329_2R.png[1] PNG 3616x2988 3616x2988+0+0 8-bit sRGB 256c 6.0015MiB 1.400u 0:01.149
ScanImage330_1L.png[2] PNG 3561x2950 3561x2950+0+0 8-bit sRGB 256c 6.12408MiB 1.310u 0:01.073
ScanImage330_2R.png[3] PNG 3616x2988 3616x2988+0+0 8-bit sRGB 256c 6.60378MiB 1.230u 0:00.979
ScanImage331_1L.png[4] PNG 3612x2972 3612x2972+0+0 8-bit sRGB 256c 6.05544MiB 1.130u 0:00.878
ScanImage331_2R.png[5] PNG 3561x2950 3561x2950+0+0 8-bit sRGB 256c 5827230B 1.030u 0:00.776
ScanImage332_1L.png[6] PNG 3561x2950 3561x2950+0+0 8-bit sRGB 256c 5.98647MiB 0.930u 0:00.675
ScanImage332_2R.png[7] PNG 3533x2950 3533x2950+0+0 8-bit sRGB 256c 5.75602MiB 0.820u 0:00.571
ScanImage333_1L.png[8] PNG 3514x2950 3514x2950+0+0 8-bit sRGB 256c 5.41659MiB 0.660u 0:00.465
ScanImage333_2R.png[9] PNG 3514x2950 3514x2950+0+0 8-bit sRGB 256c 5.9208MiB 0.420u 0:00.293
ScanImage334_1L.png[10] PNG 3514x2950 3514x2950+0+0 8-bit sRGB 256c 6.08766MiB 0.280u 0:00.185
ScanImage334_2R.png[11] PNG 3514x2950 3514x2950+0+0 8-bit sRGB 256c 985584B 0.130u 0:00.079
user@COMP:~/idkme$ ls
ScanImage329_1L.png  ScanImage333_1L.png  ScanImage337_1L.png  ScanImage341_1L.png  ScanImage345_1L.png
ScanImage329_2R.png  ScanImage333_2R.png  ScanImage337_2R.png  ScanImage341_2R.png  ScanImage345_2R.png
ScanImage330_1L.png  ScanImage334_1L.png  ScanImage338_1L.png  ScanImage342_1L.png  ScanImage346_1L.png
ScanImage330_2R.png  ScanImage334_2R.png  ScanImage338_2R.png  ScanImage342_2R.png  ScanImage346_2R.png
ScanImage331_1L.png  ScanImage335_1L.png  ScanImage339_1L.png  ScanImage343_1L.png  ScanImage347_1L.png
ScanImage331_2R.png  ScanImage335_2R.png  ScanImage339_2R.png  ScanImage343_2R.png  ScanImage347_2R.png
ScanImage332_1L.png  ScanImage336_1L.png  ScanImage340_1L.png  ScanImage344_1L.png
ScanImage332_2R.png  ScanImage336_2R.png  ScanImage340_2R.png  ScanImage344_2R.png
```
(tl;dr = Pretends like there's just 11 files, despite there being far more than 11 files)

> What about `convert *.png -verbose info:`?

I can attach a text file of the output, but suffice it to say there's not much of interest, because it too only churns through the first 11 images.

One line that sticks out, though, is:

`Scene: xx of 12`

This increments by 1 with each image, and stops at `Scene: 11 of 12`.

**Reply by YutMarma** [reply]
Here's what happens when I run `convert ScanImage335_1L.png -verbose info: > 12thPNG.txt` (ScanImage335_1L.png being the 12th .PNG, the next one in line when `convert` stops)

[12thPNG.txt](https://github.com/ImageMagick/ImageMagick/files/14610880/12thPNG.txt)

**Comment by snibgo** [comment]
> One line that sticks out, though, is: `Scene: xx of 12`

That is expected. After IM has read all the input files, we have images (aka "scenes") in a list of images. They are numbered from zero.

A filename with wildcards eg '*' is expanded into a list of files either by the shell (eg see https://www.gnu.org/software/bash/manual/html_node/Filename-Expansion.html), or within IM. I don't know the rules about this, or how to figure out which is doing the expansion.
